[WIP] Adding @eldad-a's ridge directed ring detector

alexdesiqueira opened the pull request:
## Description

Continues #1706, by @eldad-a:

> an implementation of the algorithm in
> Sci. Rep. 5, 13584; doi: 10.1038/srep13584 (2015).

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

eldad-a committed:
added ridge-directed-ring-detector + demo notebook

an implementation of the algorithm in
Sci. Rep. 5, 13584; doi: 10.1038/srep13584 (2015).

EA
 On branch ridge-directed-ring-detector
 Changes to be committed:
	new file:   ridge-directed-ring-detector/demo.ipynb
	new file:   ridge-directed-ring-detector/ridge_directed_ring_detector.pyx

stefanv committed:
Add ring detector to build system

eldad-a committed:
update ring detector

            1. Correct outer image frame error (`IndexError`)
            2. Commented out `MaxRads`, a variable not being used in the code so far
            3. Introduced a `rate` variable for future modification
               (growth rate of `ksigma` with the radius `R`)

eldad-a committed:
added data file for the ring_detector demo notebook and modified acc.

 On branch ridge-directed-ring-detector
 Your branch is up to date with 'origin/ridge-directed-ring-detector'.

 Changes to be committed:
	modified:   ridge-directed-ring-detector/demo.ipynb
	new file:   ridge-directed-ring-detector/unprocessed_fig_46.png

Commit:
[WIP] Adding ring_detector.py and its prelminary structure.

pep8speaks commented:
Hello @alexdesiqueira! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/transform/ring_detector.py`](https://github.com/scikit-image/scikit-image/blob/ac6e68f0f97aa262b46bc9f49bfa9c76a29c1bc5/skimage/transform/ring_detector.py):

> [Line 15:80](https://github.com/scikit-image/scikit-image/blob/ac6e68f0f97aa262b46bc9f49bfa9c76a29c1bc5/skimage/transform/ring_detector.py#L15): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (137 > 79 characters)
> [Line 30:26](https://github.com/scikit-image/scikit-image/blob/ac6e68f0f97aa262b46bc9f49bfa9c76a29c1bc5/skimage/transform/ring_detector.py#L30): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

alexdesiqueira commented:
The following is based on the discussion I had with @clementkng on Slack during the SciPy 2020 sprint, and what I think we need to pursue for this PR.

What we have:
```
skimage/transform/ridge-directed-ring-detector/demo.ipynb
skimage/transform/ridge-directed-ring-detector/unprocessed_fig_46.png
skimage/transform/_ring_detector.pyx
```
I'd like to:
**[OK]** 1. create a function within transform — let's say `ring_detector.py` — that would have the API necessary to call `ring_detector.pyx`. Something on the line of `skimage/transform/hough_transform.py` or `skimage/transform/radon_transform.py`;
2. move `demo.ipynb` to `doc/examples`. The notebook would be converted to a `.py` file — as the other files in our documentation. It'd go to `doc/examples/transform/`. A possible name would be `plot_ring_detector.py`.
3. remove the `ridge-directed-ring-detector` folder.
4. replace OpenCV functions on `_ring_detector.pyx` with skimage ones.

clementkng committed:
DOC: Refactor ridge directed ring detector example

eldad-a commented:
Hello @alexdesiqueira

Many thanks for moving forward with this PR, and sorry for my losing track of the process.

I should have a draft which is `opencv` independent - all spatial derivatives done by `skimage` functions

alexdesiqueira commented:
Hey @eldad-a ,
don't worry at all — it happens to all of us :smile:

> I should have a draft which is `opencv` independent - where all spatial derivatives are performed by `skimage` functions.
>
> If that's helpful, please let me know where to push it to.

That would be great, thank you! I will add you as a contributor, please feel free to push it here when you can.

mkcor commented on doc/examples/transform/plot_ring_detector.py:
```diff
@@ -0,0 +1,132 @@
+
+from skimage.transform import ring_detector
+from skimage import data
+from scipy import ndimage
```
import `ndimage` before `skimage`

mkcor commented on doc/examples/transform/plot_ring_detector.py:
```diff
@@ -0,0 +1,132 @@
+using 2d microscope imaging.
+
+Here's a quick description of the parameters. The explanations refer to the
+terms as introduced in [Afik 2015](https://www.nature.com/articles/srep13584)
```
Looks like you're using Markdown syntax. In my experience, it should be reStructuredText.

emmanuelle replied:
indeed :-)

emmanuelle replied:
you can take a look at existing gallery scripts for examples of rst syntax.

alexdesiqueira replied:
Sorry about that @mkcor @emmanuelle, but this is too raw for a review yet. :smile: I asked @clementkng to move the code from a notebook to this file, starting the documentation process. We'll take care of it :slightly_smiling_face:

mkcor commented on skimage/transform/ring_detector.py:
```diff
@@ -0,0 +1,30 @@
+from ._ring_detector import RidgeHoughTransform
+
+import matplotlib.pyplot as plt
```
Please swap import order.
Reference: https://www.python.org/dev/peps/pep-0008/#imports

mkcor commented on doc/examples/transform/plot_ring_detector.py:
```diff
@@ -0,0 +1,132 @@
+Ridge Directed Ring Detection
+=============================
+
+Robust and highly performant ring detection algorithm for 3d particle tracking
```
```suggestion
Robust and highly performant ring detection algorithm for 3D particle tracking
```
from my impression on the convention used in scikit-image's docs

Commit:
Update doc/examples/transform/plot_ring_detector.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

Commit:
Update doc/examples/transform/plot_ring_detector.py

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

Commit:
Fixing according to @mkcor suggestion

alexdesiqueira commented:
Thank you @mkcor! This is still very raw, not ready for a review, though.
Could we ping you when this is somewhat acceptable? :sweat_smile: Thank you again!

mkcor commented:
@alexdesiqueira sure, no worries! Your PR title says 'WIP' indeed :)

eldad-a committed:
removed opencv dependencies from ring_detector

1. Replace OpenCV Hessian estimation by skimage ones

2. Comment-out 'eccentricity', which relies on OpenCV for EllipseFit

3. Rename, following skimage conventions:
Lxx -> Lrr , Lyy -> Lcc , Lxy -> Lrc

3. Add performance observations and TODO to consider

eldad-a committed:
adjust 'curv_thresh' following the transition from OpenCV to skimage

eldad-a commented:
> The following is based on the discussion I had with @clementkng on Slack during the SciPy 2020 sprint, and what I think we need to pursue for this PR.
> 
> What we have:
...
> 4. replace OpenCV functions on `_ring_detector.pyx` with skimage ones.


Thank you all for your patience with me!
Hope everyone is doing well during these different times.

1. OpenCV dependency removed; currently replaced by `skimage.feature.hessian_matrix`.
2. Updated `curve_thresh` parameter in the example `plot_ring_detector.py` accordingly.

Noticed that `ring_detector.py` returns the `rings` rather than the `subpixel rings`. 
Where should the alternative be highlighted for users?
I would support having the `subpixel rings` as the default.

I've been exploring the outcome on the use-case for which this code was developed originally (an example was provided in  `ridge-directed-ring-detector/unprocessed_fig_46.png` ; BTW, was this file removed?).
Included some observations and suggestions in the `_ring_detector.pyx` itself.

I include these here as well for convenience; happy to hear your take on thees!

#### EA NOTES

* The original `opencv` version shows better performance compared to the current `skimage`:
   * `preprocessing` takes x3 longer using `hessian_matrix` (from `skimage`), and  
   * `ring_detection` takes about x1.2, showing higher sensitivity to choice of parameter (`sigma` and `curv_thresh`)
* The results seem more precise and less sensitive to choice of parameters when derivatives are estimated using the           `filters.farid_v/h` (or `filters.scharr_v/h`); however, these operators take x6  (x3) times longer to compute, w.r.t to `hessian_matrix`, which uses `np.gradient`; this could probably be improved by programming the more precise (but slower) operators to allow the 2nd order derivatives directly.

#### EA TO-Consider 

* store both principal curvatures and include verifying that |Lpp|>|Lqq| ; here Lpp denotes the least principal curvature (Lqq the largest principal curvature).
* allow the user to choose between np.gradient (skimage `hessian_matrix`), `scharr`, and `farid` operators; the latter seem to offer more precise results, with less sensitivity to the choice of parameters. The former is faster.
* add automated choice of `curv_thresh` based on percentiles.
* verify results and efficiency are not significantly affected the transition from `opencv` to `skimage` (Hessian calculation)
* Prepare PR for `directed ridge detector` ?  There are several related algorithms in `skimage`; does any of them use non-maximum suppression in the direction of the eigen-vector associated with the least-principal curvature? (as performed in the ring_detection ). Explore the effect of separating the ridge detection function entirely; it may not affect the `%%timeit` results.

alexdesiqueira commented:
Hey @eldad-a,

> 1. OpenCV dependency removed; currently replaced by `skimage.feature.hessian_matrix`.
> 2. Updated `curve_thresh` parameter in the example `plot_ring_detector.py` accordingly.

Thank you very much, we appreciate your help!

> Noticed that ring_detector.py returns the rings rather than the subpixel rings.
> Where should the alternative be highlighted for users?
> I would support having the subpixel rings as the default.

I'd ask your help to determine that technical features, Eldad. I still don't understand the algorithm that much, so please feel free to check what you consider doable on that subject — I would take a time to understand :slightly_smiling_face: 

> I've been exploring the outcome on the use-case for which this code was developed originally (an example was provided in ridge-directed-ring-detector/unprocessed_fig_46.png ; BTW, was this file removed?).

I'd like to structure the example on the lines of the examples we have already — hence `doc/examples/transform/plot_ring_detector.py`. Would you like to base the example on that specific file? If yes, I'd suggest to check the license for that file. If it's permissible, we could add it to `data` and start the example from there; what do you think?

> - allow the user to choose between np.gradient (skimage hessian_matrix), scharr, and farid operators; the latter seem to offer more precise results, with less sensitivity to the choice of parameters. The former is faster.

I like this! I'd need to understand the other points better, though; what would be their implications, you know... I need to read your paper again :slightly_smiling_face: 
Thanks again!

clementkng committed:
DOC: Fix import order and convert Markdown to RST

eldad-a commented:
Hi @alexdesiqueira :-)

 
> > 1. OpenCV dependency removed; currently replaced by `skimage.feature.hessian_matrix`.
> > 2. Updated `curve_thresh` parameter in the example `plot_ring_detector.py` accordingly.
> 
> Thank you very much, we appreciate your help!
> 

It is my pleasure;
I thank you for your support and encouragement!

> > Noticed that ring_detector.py returns the rings rather than the subpixel rings.
> > Where should the alternative be highlighted for users?
> > I would support having the subpixel rings as the default.
> 
> I'd ask your help to determine that technical features, Eldad. I still don't understand the algorithm that much, so please feel free to check what you consider doable on that subject — I would take a time to understand slightly_smiling_face

Sure, I will.

> 
> > I've been exploring the outcome on the use-case for which this code was developed originally (an example was provided in ridge-directed-ring-detector/unprocessed_fig_46.png ; BTW, was this file removed?).
> 
> I'd like to structure the example on the lines of the examples we have already — hence `doc/examples/transform/plot_ring_detector.py`. Would you like to base the example on that specific file? If yes, I'd suggest to check the license for that file. If it's permissible, we could add it to `data` and start the example from there; what do you think?

Which data image to use... Why not both?
#3384 calls for scientific examples, so this is an opportunity.

My understanding of licensing goes a very short distance (well, maybe none...).
Here's what the original publication reads in terms of license: [Rights and permissions](https://www.nature.com/articles/srep13584#rightslink).
The figure is part of [Supplementary Figure S2(a)](https://static-content.springer.com/esm/art%3A10.1038%2Fsrep13584/MediaObjects/41598_2015_BFsrep13584_MOESM1_ESM.pdf) 
in Afik, E. [Robust and highly performant ring detection algorithm for 3d particle tracking using 2d microscope imaging](https://www.nature.com/articles/srep13584). Sci. Rep. 5, 13584; doi: 10.1038/srep13584 (2015).

> > * allow the user to choose between np.gradient (skimage hessian_matrix), scharr, and farid operators; the latter seem to offer more precise results, with less sensitivity to the choice of parameters. The former is faster.
> 
> I like this! I'd need to understand the other points better, though; what would be their implications, you know... I need to read your paper again 

If you have questions, please let me know; may be easier than going over the paper.

Below there's  a TODO list I created for my future reference; some are added features, so I guess aren't a must. 
Could you share what are the priorities to have this PR ready?

### EA TODO:

* [ ]  ring_detector.py : return subpixel precision ring parameters
* [ ]  allow the user to choose between np.gradient (skimage hessian_matrix), scharr, and farid operators; the latter seem to offer more precise results, with less sensitivity to the choice of parameters. The former is faster.
* [ ]  test: store both principal curvatures and include verifying that |Lpp|>|Lqq| ; here Lpp denotes the least principal curvature (Lqq the largest principal curvature).
* [ ] add automated choice of `curv_thresh` based on percentiles.
* [ ] verify results and efficiency are not significantly affected the transition from `opencv` to `skimage` (Hessian calculation)

In case `skimage` community has interest:
* [ ] Prepare PR for `directed ridge detector` ?  There are several related algorithms in `skimage`; does any of them use non-maximum suppression in the direction of the eigen-vector associated with the least-principal curvature? (as performed in the ring_detection ). 
   * [ ] Explore the effect of separating the ridge detection function entirely; it may not affect the `%%timeit` results.

stefanv commented:
I think this one is slipping again.

alexdesiqueira commented:
@stefanv I have this in my mind for a while, now. :slightly_smiling_face: I'll come back to that next week.

alexdesiqueira commented:
Alright, (cracking knuckles :fist_right::fist_left: )
let's get back!

@eldad-a:

> Which data image to use... Why not both?
#3384 calls for scientific examples, so this is an opportunity.

Of course :slightly_smiling_face: I'd like to see a well-structured example, though. Maybe using these images to show different aspects of the algorithm? Would that be feasible?

> My understanding of licensing goes a very short distance (well, maybe none...).
Here's what the original publication reads in terms of license: Rights and permissions.
The figure is part of Supplementary Figure S2(a)
in Afik, E. Robust and highly performant ring detection algorithm for 3d particle tracking using 2d microscope imaging. Sci. Rep. 5, 13584; doi: 10.1038/srep13584 (2015).

The link says that the license is CC-BY, Eldad; we'll be good if the repository has an attribution to the paper.

> Below there's a TODO list I created for my future reference; some are added features, so I guess aren't a must.
Could you share what are the priorities to have this PR ready?

I would be happy if we have, for this PR:
* a running version of the algorithm — the first one you presented us is great already;
* a documentation page on how to use the algorithm;
* basic tests.

We can implement the TODO's you proposed in subsequent PRs, Eldad. How does that sound?

stefanv commented:
> The link says that the license is CC-BY, Eldad; we'll be good if the repository has an attribution to the paper.

Is this for an image that lives inside skimage.data?  We only add CC0 and public domain images there, I think, since we cannot easily have our users display the correct copyright otherwise.

alexdesiqueira commented:
> Is this for an image that lives inside skimage.data? We only add CC0 and public domain images there, I think, since we cannot easily have our users display the correct copyright otherwise.

Oh, sorry @stefanv; you're right! Do you have any images you didn't use in that paper — and you would consider donating — by any chance, @eldad-a?

alexdesiqueira commented:
Also, @scikit-image/core may I ask some help on "beautifying" the cython code here? Thank you!

seberg commented on skimage/transform/ring_detector.py:
```diff
@@ -0,0 +1,27 @@
+from ._ring_detector import RidgeHoughTransform
+
+__all__ = ['ring_detector']
```
The all doesn't include `RidgeHoughTransform`, but the example uses it directly.  If it is supposed to be exposed, I think the `params` need to be hidden away/defined with the cdef.  You could make them `cdef public ...` to allow the user to modify the parameters (although I wonder a bit if the parameters shouldn't just be arguments to the `ridge_hough.rings_detection` call).  As to the `img_preprocess` call, it could probably be included into the `__cinit__`.

seberg commented on skimage/transform/_ring_detector.pyx:
```diff
@@ -0,0 +1,1184 @@
+RIJ = np.uint32
+ctypedef np.uint32_t RIJ_t
+
+INDX = np.int
+ctypedef Py_ssize_t INDX_t
```
This seems incorrect on 64bit windows.  Maybe just use `IDX = np.intp` and `ctypedef npy_intp INDX_T`.  Those don't have to match, bt it is what numpy uses (I am not sure cython uses it, but...)

seberg commented on skimage/transform/_ring_detector.pyx:
```diff
@@ -0,0 +1,1184 @@
+INDX = np.int
+ctypedef Py_ssize_t INDX_t
+
+ctypedef unsigned short KINT_t
```
This isn't even used?

seberg commented on skimage/transform/_ring_detector.pyx:
```diff
@@ -0,0 +1,1184 @@
+# (assersion)
+# TODO: add edge detection version
+
+from __future__ import division 
```
Delete, just use `# cython: language_level=3` at the very top.

seberg commented on skimage/transform/_ring_detector.pyx:
```diff
@@ -0,0 +1,1184 @@
+#@cython.boundscheck(True)
+#@cython.wraparound(True)
+@cython.cdivision(True)
+cpdef votes2rings(RIJ_t [:] votes,
```
Maybe just `cdef` here and for the other helpers. Then this is completely gone from python and you don't have to worry about making sure the parameters are well defined. (because the callers already checked them).

seberg reviewed (commented):
Sorry, this is super random.  High level, I think mostly that the parameter handling should maybe be hidden away (either into the init or as function parameters to the second function). And some of the others, like the derivatives should probably just be `cdef class` attributes, so nobody can mess with them.

Commit:
Starting to un-OO _ring_detector

Commit:
Rewriting img_preprocess -> preprocessing

Commit:
RidgeHoughTransform -> hough_transform_ridge

Commit:
Deleting unused: least_principal_curvature

Commit:
Trimming comments, xrange -> range

Commit:
Adding spaces for readability, removing comments

Commit:
Removing debugging function

Commit:
DTYPE_t -> np_floats

Commit:
pi -> M_PI

Commit:
RIJ, DTYPE, INDX -> np_uints, np_ints, np_floats

Commit:
Fixing 'Fused types not allowed here'

Commit:
Fixing cnp_float, int, uint -> double, cnp.uint32

Commit:
Move ring_detector -> hough_..._ridge; remove .py file

eldad-a commented:
@alexdesiqueira 

Could you please guide me through the following:

1. Where should example images be added to (folder)?
2. Attempting to compile commit `cbab232` using the command (found at _#installation-from-source_ ):
```bash
python setup.py build_ext -i
``` 
results in :
```python
ValueError: '.../scikit-image/skimage/transform/_ring_detector.pyx' doesn't match any files
```
Any suggestions? Should I test using my original PR and commit the test files here?

alexdesiqueira commented:
Hey @eldad-a,
sorry about the silence.
> 1. Where should example images be added to (folder)?

It would be best to add these in our [GitLab repo](https://gitlab.com/scikit-image/data).

> 2. Attempting to compile commit cbab232 using the command...
> Any suggestions? Should I test using my original PR and commit the test files here?

Sorry about that, Eldad :sweat_smile: I broke the PR while converting from procedural to OO, and I didn't have time to fix it properly yet. We need to think on an API that could agree with the other transforms, so we can write the tests on that scheme. Minimally, what do you think that the function would need as input and output?
Thank you!

eldad-a commented:
@alexdesiqueira thanks!

> We need to think on an API that could agree with the other transforms, so
we can write the tests on that scheme.

Could you suggest your favorite transform on `skimage`, so I'll have a
better idea? Would skimage circle_hough transform be a good reference?




On Wed, Jun 16, 2021, 09:28 Alexandre de Siqueira ***@***.***>
wrote:

> Hey @eldad-a <https://github.com/eldad-a>,
> sorry about the silence.
>
>
>    1. Where should example images be added to (folder)?
>
> It would be best to add these in our GitLab repo
> <https://gitlab.com/scikit-image/data>.
>
>
>    1. Attempting to compile commit cbab232
>    <https://github.com/scikit-image/scikit-image/commit/cbab2322694ab1db01d73f90e06a8b4e76b58f77>
>    using the command...
>    Any suggestions? Should I test using my original PR and commit the
>    test files here?
>
> Sorry about that, Eldad 😅 I broke the PR while converting from
> procedural to OO, and I didn't have time to fix it properly yet. We need to
> think on an API that could agree with the other transforms, so we can write
> the tests on that scheme. Minimally, what do you think that the function
> would need as input and output?
> Thank you!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/4847#issuecomment-862525850>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5OFP63QBOM23CHIMV2WADTTDGKZANCNFSM4OX72ZJA>
> .
>

alexdesiqueira commented:
Hey @eldad-a,
 > Would skimage circle_hough transform be a good reference?

Sure! I think all our `hough_` transforms would be great for comparison.
Thank you!

grlee77 commented:
@alexdesiqueira, @eldad-a. I would like to release 0.19 fairly soon. Is this something you still want to try and get into that release or should we bump back the milestone here? (given the [WIP] status, I haven't tried to review anything here yet)

alexdesiqueira commented:
Hey @grlee77, I'm sorry. I think this will take a while — I removed the milestone for now.

eldad-a commented:
> Hey @eldad-a,
> sorry about the silence.
> 
> > 1. Where should example images be added to (folder)?
> 
> It would be best to add these in our [GitLab repo](https://gitlab.com/scikit-image/data).
> 
...

Hi @alexdesiqueira  ,

It seems I do not have permissions to push to the [GitLab repo](https://gitlab.com/scikit-image/data).
Who should I contact on this matter?

Other than the example images and the tests we've discussed, what is needed to complete a first round of this PR?

Many thanks!
Eldad

mkcor commented:
Hi @eldad-a,

> It seems I do not have permissions to push to the [GitLab repo](https://gitlab.com/scikit-image/data).
> Who should I contact on this matter?

Please proceed with a regular pull request, including your data file(s) and corresponding metadata: for example, see https://gitlab.com/scikit-image/data/-/merge_requests/11/diffs

eldad-a commented:
Hi @mkcor 
 
> Please proceed with a regular pull request, including your data file(s) and corresponding metadata: for example, see https://gitlab.com/scikit-image/data/-/merge_requests/11/diffs

Thanks, just did.
Hope I did it as needed.

Thanking you,
Eldad

alexdesiqueira commented:
Hey @eldad-a,
> Other than the example images and the tests we've discussed, what is needed to complete a first round of this PR?

I would like to have this function as a procedural one, not an object. My cython is not the best, and I broke most of the code so far :joy: Could you help me with that?

eldad-a commented:
Hi @alexdesiqueira ,

Sure would you like to set a time to meet and discuss?

mkcor commented:
PR https://gitlab.com/scikit-image/data/-/merge_requests/13 by @eldad-a is now ready :slightly_smiling_face:

alexdesiqueira commented:
Happy New Year everyone! This PR (if we consider also its parent, #1706) is now more than 6 years old. :birthday::partying_face:
I'll take care of it this month; I promise! :sweat_smile:

alexdesiqueira commented:
> I'll take care of it this month; I promise! :sweat_smile:

I was informed that this is too bold of a statement: let me rephrase this as _I'll start taking care of it this month_ :joy:

Commit:
Removing a small todo to check if my copy still works

Commit:
Add test for central moment computation with spacing and fix implementation.

Commit:
Add moments test with spacing.

rfezzani committed:
Fix first pass declaration type

Commit:
colour -> color on examples/transform/plot_ring_detector

Commit:
dr -> delta_r, moving hough_ridge to hough_transform

Commit:
hough_ridge -> hough_transform

Commit:
_ring_detector -> _hough_transform_ridge on setup

Commit:
Divides img_preproc functions, returns values in directed_ridge_detector

Commit:
Refactor function names

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
